rotkehlchen/balances/historical.py

```diff
@@ -0,0 +1,113 @@
+from collections import defaultdict
+from typing import TYPE_CHECKING, TypedDict
+
+from rotkehlchen.assets.asset import Asset
+from rotkehlchen.constants.prices import ZERO_PRICE
+from rotkehlchen.db.filtering import HistoryEventFilterQuery
+from rotkehlchen.db.history_events import DBHistoryEvents
+from rotkehlchen.errors.misc import InputError, RemoteError
+from rotkehlchen.errors.price import NoPriceForGivenTimestamp
+from rotkehlchen.fval import FVal
+from rotkehlchen.history.events.structures.types import EventDirection
+from rotkehlchen.history.price import PriceHistorian
+from rotkehlchen.types import Timestamp
+
+if TYPE_CHECKING:
+    from rotkehlchen.db.dbhandler import DBHandler
+
+
+class HistoricalBalance(TypedDict):
+    amount: FVal
+    price: FVal
+
+
+class HistoricalBalancesManager:
+
+    def __init__(self, db: 'DBHandler') -> None:
+        self.db = db
+
+    def get_balance(self, timestamp: Timestamp) -> dict[Asset, HistoricalBalance]:
+        """Get historical balances for all assets at a given timestamp
+
+        May raise:
+        - InputError if balance for the given timestamp does not exist.
+        """
+        events, main_currency = self._get_events_and_currency(timestamp)
+        if len(events) == 0:
+            raise InputError('No historical data found for given timestamp.')
+
+        amounts = self._calculate_balance_from_events(events)
+        result: dict[Asset, HistoricalBalance] = {}
+        for asset, amount in amounts.items():
+            try:
+                price = PriceHistorian().query_historical_price(
+                    from_asset=asset,
+                    to_asset=main_currency,
+                    timestamp=timestamp,
+                )
+            except (RemoteError, NoPriceForGivenTimestamp):
+                price = ZERO_PRICE
+
+            result[asset] = {'amount': amount, 'price': price}
+
+        return result
+
+    def get_asset_balance(self, asset: Asset, timestamp: Timestamp) -> HistoricalBalance:
+        """Get historical balance for a single asset at a given timestamp
+
+        May raise:
+        - InputError if balance for the asset at the given timestamp does not exist.
+        """
+        events, main_currency = self._get_events_and_currency(timestamp, (asset,))
+        if len(events) == 0:
+            raise InputError(f'No historical data found for {asset} at given timestamp.')
```

Comment:
How about introducing a NotFoundError? A new exception? As this is not really an input error. There is just no events present for the asset until the given timestamp

```suggestion
            raise NotFoundError(f'No historical data found for {asset} until the given timestamp.')
```